pandas/core/indexes/period.py

```diff
@@ -165,90 +203,205 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
+        self._data._freq = value
+
+    def _shallow_copy(self, values=None, **kwargs):
+        # TODO: simplify, figure out type of values
+        if values is None:
+            values = self._data
+
+        if isinstance(values, type(self)):
+            values = values._values
+
+        if not isinstance(values, PeriodArray):
+            if (isinstance(values, np.ndarray) and
+                    is_integer_dtype(values.dtype)):
+                values = PeriodArray(values, freq=self.freq)
+            else:
+                # in particular, I would like to avoid period_array here.
+                # Some people seem to be calling use with unexpected types
+                # Index.difference -> ndarray[Period]
+                # DatetimelikeIndexOpsMixin.repeat -> ndarray[ordinal]
+                # I think that once all of Datetime* are EAs, we can simplify
+                # this quite a bit.
+                values = period_array(values, freq=self.freq)
+
+        # I don't like overloading shallow_copy with freq changes.
+        # See if it's used anywhere outside of test_resample_empty_dataframe
+        attributes = self._get_attributes_dict()
+        freq = kwargs.pop("freq", None)
+        if freq:
+            values = values.asfreq(freq)
+            attributes.pop("freq", None)
+
+        attributes.update(kwargs)
+        if not len(values) and 'dtype' not in kwargs:
+            attributes['dtype'] = self.dtype
+        return self._simple_new(values, **attributes)
+
+    def _shallow_copy_with_infer(self, values=None, **kwargs):
         """ we always want to return a PeriodIndex """
         return self._shallow_copy(values=values, **kwargs)
 
-    def _coerce_scalar_to_index(self, item):
-        """
-        we need to coerce a scalar to a compat for our index type
+    # ------------------------------------------------------------------------
+    # Boxing
+    # err. maybe not...
 
-        Parameters
-        ----------
-        item : scalar item to coerce
+    @property
+    def _box_func(self):
+        return self._data._box_func
+
+    def _maybe_box_as_values(self, values, **attribs):
+        """Box an array of ordinals to a PeriodArray
+
+        This is purely for compatibility between PeriodIndex
+        and Datetime/TimedeltaIndex. Once these are all backed by
+        an ExtensionArray, this can be removed
         """
-        return PeriodIndex([item], **self._get_attributes_dict())
+        # TODO(DatetimeArray): remove
+        freq = attribs['freq']
+        return PeriodArray(values, freq=freq)
+
+    # ------------------------------------------------------------------------
+    # Dispatch and maybe box. Not done in delegate_names because we box
+    # different from those (which use Index).
+
+    def asfreq(self, freq=None, how='E'):
+        result = self._data.asfreq(freq=freq, how=how)
+        return self._simple_new(result, name=self.name)
+
+    def _nat_new(self, box=True):
```

Comment:
is this not in a super class?

Reply:
Will wait on https://github.com/pandas-dev/pandas/pull/23215#discussion_r226662611